libraries/app/api.cpp

```diff
@@ -304,23 +304,25 @@ namespace graphene { namespace app {
        vector<operation_history_object> result;
        const auto& stats = account(db).statistics(db);
        if( stats.most_recent_op == account_transaction_history_id_type() ) return result;
+       if( start.instance.value < stop.instance.value) return result;
+       if( start.instance.value == stop.instance.value && start != operation_history_id_type()) return result;
        const account_transaction_history_object* node = &stats.most_recent_op(db);
        if( start == operation_history_id_type() )
           start = node->operation_id;
+       if( start > node->operation_id )
+          start = node->operation_id;
 
-       while(node && node->operation_id.instance.value > stop.instance.value && result.size() < limit)
-       {
-          if( node->operation_id.instance.value <= start.instance.value )
-             result.push_back( node->operation_id(db) );
-          if( node->next == account_transaction_history_id_type() )
-             node = nullptr;
-          else node = &node->next(db);
-       }
-       if( stop.instance.value == 0 && result.size() < limit )
+       const auto& hist_idx = db.get_index_type<account_transaction_history_index>();
+       const auto& by_op_idx = hist_idx.indices().get<by_op>();
+       auto index_end = by_op_idx.end();
+       auto itr = by_op_idx.lower_bound(boost::make_tuple(account, start));
+
+       while(itr != index_end && result.size() < limit)
```

Comment:
There is still something wrong here.

We really need a unit test for this feature. Try write one? I think you will do better after done it.

The `operation_history_id_type()` is ambitious. However, we can define it as:
* if the account has `operation_history_id_type()` (aka `1.11.0`), **and** if `stop` is it, **and** if didn't reach limit, return it;
* otherwise, don't return `stop`.

For example, if we have data:
* account A with records `{ 5, 3, 1, 0 }`, and
* account B with records `{ 6, 4, 2 }`,

Expected results:
* f(A, 0, 4, 9) = { 5, 3, 1, 0 }
* f(A, 0, 4, 6) = { 5, 3, 1, 0 }
* f(A, 0, 4, 5) = { 5, 3, 1, 0 }
* f(A, 0, 4, 4) = { 3, 1, 0 }
* f(A, 0, 4, 3) = { 3, 1, 0 }
* f(A, 0, 4, 2) = { 1, 0 }
* f(A, 0, 4, 1) = { 1, 0 }
* f(A, 0, 4, 0) = { 5, 3, 1, 0 }
* f(A, 1, 5, 9) = { 5, 3 }
* f(A, 1, 5, 6) = { 5, 3 }
* f(A, 1, 5, 5) = { 5, 3 }
* f(A, 1, 5, 4) = { 3 }
* f(A, 1, 5, 3) = { 3 }
* f(A, 1, 5, 2) = { }
* f(A, 1, 5, 1) = { }
* f(A, 1, 5, 0) = { 5, 3 }
* f(A, 0, 3, 9) = { 5, 3, 1 }
* f(A, 0, 3, 6) = { 5, 3, 1 }
* f(A, 0, 3, 5) = { 5, 3, 1 }
* f(A, 0, 3, 4) = { 3, 1, 0 }
* f(A, 0, 3, 3) = { 3, 1, 0 }
* f(A, 0, 3, 2) = { 1, 0 }
* f(A, 0, 3, 1) = { 1, 0 }
* f(A, 0, 3, 0) = { 5, 3, 1 }
* f(B, 0, 4, 9) = { 6, 4, 2 }
* f(B, 0, 4, 6) = { 6, 4, 2 }
* f(B, 0, 4, 5) = { 4, 2 }
* f(B, 0, 4, 4) = { 4, 2 }
* f(B, 0, 4, 3) = { 2 }
* f(B, 0, 4, 2) = { 2 }
* f(B, 0, 4, 1) = { }
* f(B, 0, 4, 0) = { 6, 4, 2 }
* f(B, 2, 4, 9) = { 6, 4 }
* f(B, 2, 4, 6) = { 6, 4 }
* f(B, 2, 4, 5) = { 4 }
* f(B, 2, 4, 4) = { 4 }
* f(B, 2, 4, 3) = { }
* f(B, 2, 4, 2) = { }
* f(B, 2, 4, 1) = { }
* f(B, 2, 4, 0) = { 6, 4 }

With current code, quite some cases listed above won't pass.

And we can have more test cases, for example, another account C has no history, another account D doesn't exist, if `limit` is `0` always return an empty set, etc.

Reply:
will do the test cases, thanks again.